src/string.cr

```diff
@@ -903,23 +905,23 @@ class String
     end
   end
 
-  def byte_slice(start : Int)
+  def byte_slice(start : Int) : String
     byte_slice start, bytesize - start
   end
 
-  def codepoint_at(index)
+  def codepoint_at(index) : Int
     char_at(index).ord
   end
 
-  def byte_at(index)
+  def byte_at(index) : Int
     byte_at(index) { raise IndexError.new }
   end
 
-  def byte_at?(index)
+  def byte_at?(index) : UInt8?
     byte_at(index) { nil }
   end
 
-  def byte_at(index)
+  def byte_at(index, &block : -> U) : UInt8 | U forall U
```

Comment:
this doesn't work, because of an overload of `byte_at`:

```crystal

  def byte_at?(index) : UInt8?
    byte_at(index) { nil }
  end
```
which returns `UInt8?` (so the `U` here is `Nil`).